### PR TITLE
Dynamics2

### PR DESCRIPTION
**Pull request by 22aronl**
Race Simulation Rewrite

### CODE REVIEW

File: battery.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+
+class Battery:
+    
+    def __init__(self, max_capacity, current_capacity = -1):
+        if(current_capacity == -1):
```
Comment (JimothyGreene) [inline battery.py]: not a functional thing, but I would usually put the default value here to `None` instead of -1

File: car.py [hunk 2]
```diff
@@ -1,129 +1,50 @@
...
-        self.voltageOfaSingeCell = voltageOfaSingeCell
-        self.capacity = capacity
-        self.totalCapacity = capacity
-        self.weather_conditions = weather.Weather(latitude,longitude, startingTime)
-        self.elec = electronics.Electronics()
-        self.mot = motor.Motor()
-        self.prevTime = startingTime
-        self.deltaTime = 0
+    def __init__(self, track: Track, time: int) -> None:
+        max_capacity = 5400 # Wh
```
Comment (JimothyGreene) [inline car.py]: Make this a constant up top. i.e. something like `MAX_BATTERY_CAPACITY_WH = 5400`, and then `max_capacity = MAX_BATTERY_CAPACITY_WH`

File: car.py [hunk 3]
```diff
@@ -1,129 +1,50 @@
...
-        elif battery > 100:
-            battery = 100
-
-        print('Power draw from motor: {0:.2f} W'.format(motor_power))
-        print('Power provided from array: {0:.2f} W'.format(array_power))
-        print('Power draw from electronics {0:.2f} W'.format(electronics_power))
-        print('Battery is at {0:.2f}%'.format(battery))
-        print('Battery capacity is {0:.2f}% W', format(self.capacity))
+        arrayPower = self.array.get_power()
+        electronicsPower = 77 #power draw from constantly used systems
```
Comment (JimothyGreene) [inline car.py]: again make this a defined constant. in general, avoid "magic numbers"

Comment (JimothyGreene) [reply car.py]: also if this isn't actually going to change during runtime, you don't need to make it a variable at all

File: car.py [hunk 4]
```diff
@@ -1,129 +1,50 @@
...
-
-
-#main function to test the simulator for one set of inputs
-#need main module to run race strategy and set parameters from tack
-def main():
-    speed = 0 #m/s
-    acc = 4 #m/s^2
-    targetSpeed = 13 #m/s
+        self.motor.updateParameters(speed, slope, time)
+        motorPower = self.motor.getCurrent() * 96 #no clue what the voltage of the battery is
```
Comment (JimothyGreene) [inline car.py]: magic number

File: car.py [hunk 5]
```diff
@@ -1,129 +1,50 @@
...
+        if motorPower < 0:
+            motorPower = 0
```
Comment (JimothyGreene) [inline car.py]: this can all sorta be combined like this: `motorPower = max(0, self.motor.getCurrent() * 96)`

File: car.py [hunk 6]
```diff
@@ -1,129 +1,50 @@
...
+        return speed * deltaTime * np.cos(slope)
+        
+if __name__ == "__main__":
+    t = Track("track2.json")
+    car = Car(t, datetime.datetime(2023, 1, 10, 9, 0, 0).timestamp())
+    i = 0
+    while True:
+        car.drive(10, 0, i)
+        i += 1
+        time.sleep(1)
```
Comment (JimothyGreene) [inline car.py:54]: is this just a tester thing?

File: old/aaron_array.py [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
-from weather import Weather
+from old.weather import Weather
 
 
 #model for a solar cell array
 class Array:
     def __init__(self, weather, temperature=25):
         self.num_c60 = 158 #temporary number of cells
-        self.num_e60 = 88
+        self.num_e60 =  88
```
Comment (JimothyGreene) [inline old/aaron_array.py]: whitespace can make python weird

File: old/aaron_array.py [hunk 8]
```diff
@@ -1,13 +1,13 @@
...
         self.solarcell_c60 = SolarCell(153.328, 0.225, temperature, 0.00342) #unsure of the temperature_coefficients
         self.solarcell_e60 = SolarCell(153.328, 0.237, temperature, 0.00363)
```
Comment (JimothyGreene) [inline old/aaron_array.py]: magic nummmmmmmmmmmmbers

File: sim.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import datetime, time, math
+
+from car import Car
+from track import Track
+
+trackFile = "track2.json"
```
Comment (JimothyGreene) [inline sim.py]: might be nice to make this file take command line args with the file JSON file

File: solar_array.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+class Array:
+    def __init__(self, car, temperature):
+        self.weather = weather.Weather(car.lat, car.lon, car.time)
+        self.temperature = temperature# TODO: self.weather.get_temperature() #add method to weather
+        
+        self.num_c60 = 158 #temporary number of cells
+        self.num_e60 =  88
+        self.solarcell_c60 = SolarCell(153.328, 0.225, self.temperature, 0.00342) #unsure of the temperature_coefficients
+        self.solarcell_e60 = SolarCell(153.328, 0.237, self.temperature, 0.00363)
```
Comment (JimothyGreene) [inline solar_array.py]: numeros de magia

File: track.py [hunk 11]
```diff
@@ -1,153 +1,87 @@
...
+    minLat, maxLat, minLon, maxLon = 90, -90, 180, -180
+    for point in points:
+        if point[0] < minLon:
+            minLon = point[0]
+        if point[0] > maxLon:
+            maxLon = point[0]
+        if point[1] < minLat:
+            minLat = point[1]
+        if point[1] > maxLat:
+            maxLat = point[1]
```
Comment (JimothyGreene) [inline track.py]: this can be simplified again using min/max functions

File: track.py [hunk 12]
```diff
@@ -1,153 +1,87 @@
...
+    aspectRatio = (maxLat - minLat)/(maxLon - minLon)
+
+    # Scale points and convert to km
+    for point in points:
+        point[0] = (point[0] - minLon)/(maxLon - minLon)
+        point[1] = (point[1] - minLat)/(maxLat - minLat)
+    points = np.array(points)
+    points[:,1] = points[:,1]*aspectRatio
+    points = points*.65 #rough conversion to km from lat/lon
+    # ^- conversion not accurate, maybe an issue the points. Should be .4 multiple, but the track would too short then
```
Comment (JimothyGreene) [inline track.py]: do we need to make this accurate..?

File: track.py [hunk 13]
```diff
@@ -1,153 +1,87 @@
...
+  # max speed at T according to friction
+  # math.sqrt(self.GRAVITY * self.mass * self.tire_fricts * radius)
+  def maxSpeed(self, t: float, m: float, radius: float) -> float:
+    return np.sqrt(9.8, * m * self.friction * radius)
+      
+
+if __name__ == "__main__":
+    t = Track("track2.json")
+    cmr = t.cmr
+    print(t.trackLength)
```
Comment (JimothyGreene) [inline track.py]: again i'm assuming testing?

File: util/splinePlotter.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
+"""Helper functions for plotting."""
+import matplotlib.pyplot as _plt
```
Comment (JimothyGreene) [inline util/splinePlotter.py:2]: little unsure what's going on w this file, but i'm gonna trust it lol

File: weather_data/weather.json [hunk 15]
```diff
@@ -1 +1 @@
...ntribution": 0.0}, "KUNK1": {"distance": 44238.0, "latitude": 39.051, "longitude": -95.191, "useCount": 0, "id": "KUNK1", "name": "KU-NESA KS US SCAN", "quality": 0, "contribution": 0.0}, "72450803997": {"distance": 41258.0, "latitude": 39.008, "longitude": -95.212, "useCount": 0, "id": "72450803997", "name": "LAWRENCE ASOS, KS US", "quality": 100, "contribution": 0.0}, "F7360": {"distance": 27515.0, "latitude": 39.17, "longitude": -95.624, "useCount": 0, "id": "F7360", "name": "FW7360 Topeka KS US", "quality": 0, "contribution": 0.0}, "KTOP": {"distance": 16702.0, "latitude": 39.07, "longitude": -95.62, "useCount": 0, "id": "KTOP", "name": "KTOP", "quality": 100, "contribution": 0.0}, "AV314": {"distance": 63742.0, "latitude": 38.966, "longitude": -94.942, "useCount": 0, "id": "AV314", "name": "W0QQ De Soto KS US", "quality": 0, "contribution": 0.0}, "72456013996": {"distance": 16823.0, "latitude": 39.073, "longitude": -95.626, "useCount": 0, "id": "72456013996", "name": "TOPEKA ASOS, KS US", "quality": 100, "contribution": 0.0}, "KLWC": {"distance": 40604.0, "latitude": 39.01, "longitude": -95.22, "useCount": 0, "id": "KLWC", "name": "KLWC", "quality": 100, "contribution": 0.0}}}
```
Comment (JimothyGreene) [inline weather_data/weather.json:1]: should this be tracked by git?

**Review by JimothyGreene — CHANGES_REQUESTED** [review]
A few overarching things: 
- Python standards are to use snake case for variables and functions. Looks like you sorta use a mix of things. Try to standardize to snake case
- Lots of magic numbers. Magic is difficult to understand
- I'm not really sure how any of this was tested or what the result of the testing was. When you open a PR, it's good to provide a description of what the changes introduce, how the things you did fit into the overall system, and how it was tested. Gives us a good idea of what's going on before we review it

I left more specific comments in the files

File: car.py (hunk 6 above)
Comment (kimyejean02) [reply car.py:54]: yes

File: track.py (hunk 12 above)
Comment (rohitalamgari) [reply track.py]: I think this conversion is good enough for us without losing too much accuracy

**Review by kimyejean02 — APPROVED** [review]

**Pull request merged**